src/expanduser.cpp

```diff
@@ -1,25 +1,203 @@
+        CASE_RET_STRING(CEPH_SUPER_MAGIC)
+        CASE_RET_STRING(TMPFS_MAGIC)
+        CASE_RET_STRING(OVERLAYFS_SUPER_MAGIC)
+        CASE_RET_STRING(EXT2_OLD_SUPER_MAGIC)
+    case EXT4_SUPER_MAGIC: return "EXT2/3/4_SUPER_MAGIC";
+    default: return "<Unknown magic>";
+    }
+}
+
+bool IsNetworked(unsigned long ft)
+{
+    switch(ft)
+    {
+    case NFS_SUPER_MAGIC:   // fall through
+    case SMB_SUPER_MAGIC:   // fall through
+    case SMB2_MAGIC_NUMBER: // fall through
+    case CIFS_MAGIC_NUMBER: // fall through
+    case CODA_SUPER_MAGIC:  // fall through
+    case OCFS2_SUPER_MAGIC: // fall through
+    case AFS_SUPER_MAGIC:   // fall through
+    case LL_SUPER_MAGIC:    // fall through
+    case CEPH_SUPER_MAGIC: return true;
+    default: return false;
+    }
+}
+} // namespace
+
+#undef CASE_RET_STRING
+#endif // __linux__
+
+bool IsNetworkedFilesystem(const boost::filesystem::path& path_)
 {
-    const char* home_dir = GetStringEnv(HOME{});
-    if(home_dir == nullptr || home_dir == std::string("/") || home_dir == std::string(""))
+#ifdef __linux__
+    // Non-DEV builds put user databases in ~/.config/miopen by default; the binary cache is placed
+    // in ~/.cache/miopen. If these directories do not exist, this is not a problem, because the
+    // library creates them as needed.
+    //
+    // The problem is that statfs doesn't work in this case, and we need to determine the type of FS
+    // _before_ the databases are created.
+    //
+    // Solution (A): Just create_directories if path doesn't exist. It looks like a hack: if the
+    // path is on NFS, then the library will not use it, but will create directories (which is not
+    // good). And this function should not have any side effects on the file system.
+    //
+    // Solution (B): Traverse the input path up to the first existing directory, then check that
+    // directory. Stop after some fixed number of iterations to protect against possible file system
+    // problems. Let's use this solution for now.
+    auto path = path_;
+    for(int i = 0; i < 32; ++i)
     {
-        // todo:
-        // need to figure out what is the correct thing to do here
-        // in tensoflow unit tests run via bazel, $HOME is not set, so this can happen
-        // setting home_dir to the /tmp for now
-        home_dir = "/tmp";
+        if(boost::filesystem::exists(path))
+            break;
+        MIOPEN_LOG_NQI2("Path does not exist: '" << path.string() << '\'');
+        path = path.parent_path();
+        if(path.empty())
+            break;
     }
+    struct statfs stat;
+    const int rc = statfs(path.c_str(), &stat);
+    if(rc != 0)
+    {
+        // NOLINTNEXTLINE (concurrency-mt-unsafe)
+        MIOPEN_LOG_NQE("statfs('" << path.string() << "') rc = " << rc << ", '" << strerror(errno)
+                                  << "'");
+        return false;
+    }
+    MIOPEN_LOG_NQI("Filesystem type at '" << path.string() << "' is: 0x" << std::hex << stat.f_type
```

Comment:
@JehandadKhan Some additional logging here. `FindRecordUnsafe` prints the name of the file.

### Example for find-db.

If db file resides on EXT4 FS:
```
MIOpen(HIP): Info [IsNetworkedFilesystem] Filesystem type at '/home/atamazov/nfs-home/.config/miopen/' is: 0xef53 'EXT2/3/4_SUPER_MAGIC'
...
MIOpen(HIP): Info2 [FindRecordUnsafe] Looking for key 3-32-32-3x3-32-30-30-100-0x0-1x1-1x1-0-NCHW-FP32-F in cache for file /home/atamazov/nfs-home/.config/miopen//gfx1030_36.HIP.2_20_0_d51f49756.ufdb.txt
```
NFS case (FindRecordUnsafe prints ".../tmp/..."):
```
MIOpen(HIP): Info2 [IsNetworkedFilesystem] Path does not exist: '/home/atamazov/nfs-home/.config/miopen/'
MIOpen(HIP): Info2 [IsNetworkedFilesystem] Path does not exist: '/home/atamazov/nfs-home/.config/miopen'
MIOpen(HIP): Info2 [IsNetworkedFilesystem] Path does not exist: '/home/atamazov/nfs-home/.config'
MIOpen(HIP): Info [IsNetworkedFilesystem] Filesystem type at '/home/atamazov/nfs-home' is: 0x6969 'NFS_SUPER_MAGIC'
...
MIOpen(HIP): Info2 [FindRecordUnsafe] Looking for key 3-32-32-3x3-32-30-30-100-0x0-1x1-1x1-0-NCHW-FP32-F in cache for file /tmp/gfx1030_36.HIP.2_20_0_d51f49756.ufdb.txt
```

Reply:
This detects one particular type of a network file system - NFS. There is also another popular one - [Lustre](https://en.wikipedia.org/wiki/Lustre_(file_system)) and its variants.

Reply:
@dmikushin No, it handles many filesystem types, including LUSTRE.